### PR TITLE
Fix infinite live rest custom data polling

### PR DESCRIPTION
**Pull request by Martin-Molinero**
#### Description

- Adding `IStreamReader.RateLimit { get; }` specifying if calls to
`ReadLine()` should be rate limited by the source readers.
- Adding three failling unit tests in `master`
    - Live rest custom data: returning null, throwing an exception, returning old data will not infinitely poll

#### Related Issue

Closes https://github.com/QuantConnect/Lean/issues/3156
#### Motivation and Context

Fix infinite live custom data rest polling
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

Unit and regression tests
#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Interfaces/IStreamReader.cs [hunk 1]
```diff
@@ -34,8 +34,13 @@ public interface IStreamReader : IDisposable
...
         /// <summary>
-        /// Gets the next line/batch of content from the stream 
+        /// Gets the next line/batch of content from the stream
         /// </summary>
         string ReadLine();
+
+        /// <summary>
+        /// Gets whether or not this stream reader should be rate limited
+        /// </summary>
+        bool RateLimit { get; }
```
Comment (StefanoRaggi) [inline Common/Interfaces/IStreamReader.cs]: Just my personal preference, this property name sounds better as `ShouldBeRateLimited`

File: Engine/DataFeeds/CollectionSubscriptionDataSourceReader.cs [hunk 2]
```diff
@@ -115,7 +115,10 @@ public IEnumerable<BaseData> Read(SubscriptionDataSource source)
                     catch (Exception err)
                     {
                         OnReaderError(raw, err);
-                        continue;
+                        if (!reader.RateLimit)
+                        {
+                            continue;
+                        }
```
Comment (StefanoRaggi) [inline Engine/DataFeeds/CollectionSubscriptionDataSourceReader.cs]: The current implementations of `RateLimit` only return `true` in live mode, so L130 would never be reached with `instances == null`. Currently not a bug, but just to be proactive, maybe better add a null check on `instances` before L130?

File: Tests/Engine/DataFeeds/LiveTradingDataFeedTests.cs [hunk 3]
```diff
@@ -485,6 +488,63 @@ public void HandlesManyCustomDataSubscriptions()
             Assert.IsTrue(emittedData);
         }
 
+        [TestCase(FileFormat.Csv)]
+        [TestCase(FileFormat.Collection)]
+        public void RestCustomDataReturningNullDoNotInfinitelyPoll(FileFormat fileFormat)
```
Comment (StefanoRaggi) [inline Tests/Engine/DataFeeds/LiveTradingDataFeedTests.cs]: A small typo: `DoNot` -> `DoesNot`

File: Tests/Engine/DataFeeds/LiveTradingDataFeedTests.cs [hunk 4]
```diff
@@ -967,4 +1027,65 @@ protected override ITimeProvider GetTimeProvider()
             return _timeProvider;
         }
     }
+
+    internal class TestCustomData : BaseData
```
Comment (StefanoRaggi) [inline Tests/Engine/DataFeeds/LiveTradingDataFeedTests.cs]: Nice tests! 👍

File: Common/Interfaces/IStreamReader.cs (hunk 1 above)
Comment (Martin-Molinero) [reply Common/Interfaces/IStreamReader.cs]: Your right 😄 I'll rename it

File: Tests/Engine/DataFeeds/LiveTradingDataFeedTests.cs (hunk 4 above)
Comment (Martin-Molinero) [reply Tests/Engine/DataFeeds/LiveTradingDataFeedTests.cs]: Thanks 👍

File: Engine/DataFeeds/CollectionSubscriptionDataSourceReader.cs (hunk 2 above)
Comment (Martin-Molinero) [reply Engine/DataFeeds/CollectionSubscriptionDataSourceReader.cs]: Okay, I'll add the null check just in case 👍

**Pull request merged**